articles/what-is-an-abi/index.md

```diff
@@ -0,0 +1,268 @@
+
+This turns out to be very inefficient because if we have a lot of register in play and we call a function that doesn't use any of them we would end up filling the stack for no reason. It turns out that there is a better way to do this. It is the concept of caller-saved and callee-saved registers. If a register is marked as callee-saved and the callee wants to that register then they would have to back it up (push it onto the stack) before using it and restore it (pop it from the stack) before they returning.
+
+## Passing data to the callee and back
+
+Passing data is done in two different ways. The first is if you compile a program for 32 bits (x86) then all arguments are passed on the stack. The second way is if you compile for 64 bits (x64 architecture) then registers are used to pass arguments. If you have a lot of arguments then some are put onto the stack.
+
+The reason we can use the stack to pass arguments is because the stack is predictable enough that at compile-time, offsets from the top of the stack can be used. Here is a picture to make the point more clear.
+
+![alt text](image1.png "Call stack")
+
+*`rsp` and `esp` refer to the same register. `esp` is used in an x86 architecture and is the lower 4 bytes of `rsp` which is used in an x64 architecture.*
+*Note: a lot of the time the stack is drawn upside down for whatever reason so make sure you look at the direction of increasing addresses.*
+
+The stack pointer, `rsp` (`esp` in x86), always points to the top of the stack. From there we can apply an offset known at compile time to get the arguments. This offset is calculated based on the size that the callee allocated for local variables, the numbers of callee-saved registers saved, one pointer sized value for the return address, and finally the parameter sizes and alignments or each parameter before it. Note that when a function allocates space for local variable it will allocate it insuring there is a [16-byte alinement](https://stackoverflow.com/a/49397524/9664285) (some times even 32-bytes or 64-bytes depending on SIMD hardware being used <!-- well right now more then 16 is never required --> but you can't rely on that) by adding padding so that everything that follows has a known alignment.
+
+So, let's look at how parameters are passed to the callee in C when there are registers available to use. Note, this is only the case with the architecture is x64. We figure out how each argument is passed individually based on its type-kind. We do this for each argument:
+
+- If the type-kind of the argument is an Integers then we will pick the next available general-purpose argument register (they are in order `rdi`, `rsi`, `rdx`, `rcx`, `r8`, `r9`). If there are no available registers then the argument will be placed in the next argument position on the stack.

```

Comment:
is an Integers then ->  is an Integer, then

Reply:
(they are in order -> (in order, they are

Reply:
Oops, forgot this one on the first commit